src/function/fetch.rs

```diff
@@ -105,32 +105,39 @@ where
     ) -> Option<&'db Memo<'db, C>> {
         let database_key_index = self.database_key_index(id);
         // Try to claim this query: if someone else has claimed it already, go back and start again.
-        let claim_guard = match self.sync_table.try_claim(zalsa, id, Reentrancy::Allow) {
-            ClaimResult::Claimed(guard) => guard,
-            ClaimResult::Running(blocked_on) => {
-                blocked_on.block_on(zalsa);
+        let claim_guard = loop {
+            match self
+                .sync_table
+                .try_claim(zalsa, zalsa_local, id, Reentrancy::Allow)
+            {
+                ClaimResult::Claimed(guard) => break guard,
+                ClaimResult::Running(blocked_on) => {
+                    if !blocked_on.block_on(zalsa) {
+                        continue;
+                    }
```

Comment:
Could we just return `None` in that case, the outer `fetch` loop should then retry.